greenflow/greenflow/dataframe_flow/taskGraph.py

```diff
@@ -515,59 +516,96 @@ def register_node(self, module_name, classObj):
             add_module_from_base64(module_name, encoded_class)
             self.__widget.cache = cacheCopy
 
-    def _run(self, outputs=None, replace=None, profile=False, formated=False):
+    def _run(self, outputs=None, replace=None, profile=None, formated=False,
+             build=True):
         replace = dict() if replace is None else replace
 
-        self.build(replace, profile)
+        if build:
+            self.build(replace, profile)
+        else:
+            if replace:
+                warnings.warn(
+                    'Replace is specified, but build is set to False. No '
+                    'replacement will be done. Either set build=True or '
+                    'first build with replace then call run.',
+                    RuntimeWarning)
 
-        graph_outputs = []
-        # add the output graph only if the Output Node is not in the graph
-        found_output_node = False
-        for task in self:
-            if (task[TaskSpecSchema.task_id] == OUTPUT_ID or
-                    task[TaskSpecSchema.node_type] == OUTPUT_TYPE):
-                found_output_node = True
-                outputs_collector_node = self[task[TaskSpecSchema.task_id]]
-                for input_item in outputs_collector_node.inputs:
-                    from_node_id = input_item['from_node'].uid
-                    fromStr = from_node_id+'.'+input_item['from_port']
-                    graph_outputs.append(fromStr)
-                break
+            if profile is not None:
+                warnings.warn(
+                    'Profile is specified, but build is set to False. '
+                    'Profile will be done according to last build. '
+                    'Alternatively either set build=True or first build with '
+                    'desired profile option then call run.',
+                    RuntimeWarning)
 
+            # Reset visited status to run the taskgraph. This is done during
+            # build, but since not building need to reset here.
+            for inode in self.__node_dict.values():
+                inode.visited = False
+
+        using_output_node = False
         if outputs is None:
-            outputs = graph_outputs
+            graph_outputs = []
+            outputs = graph_outputs  # reference copy
+            # Find the output collector in the task graph.
+            for task in self:
+                # FIXME: Note the inconsistency of getting a task
+                #     "for task in self" yet also retrieving a node
+                #     via "self[task_id]".
+                node = self[task[TaskSpecSchema.task_id]]
+                if node.node_type_str == OUTPUT_TYPE:
+                    using_output_node = True
```

Comment:
That variable `using_output_node` maybe should be called `using_taskgraphs_output_node`.  It's set to True when `outputs` is not specified and taskgraph has an output collector, otherwise it's false.

When `outputs` is specified, an output collector is created and later (after taskgraph run logic when results are obtained) deleted, but this new output collector is not the same output collector that a taskgraph might already have. I did not like the logic of rewiring the an existing output collector for a taskgraph (if one happened to exists). If outputs is specified, that's independent of an output collector that might be set in the task graph.

Reply:
ok